hail/src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala

```diff
@@ -49,48 +47,46 @@ object IntervalFunctions extends RegistryFunctions {
     }
 
     registerCodeWithMissingness("start", TInterval(tv("T")), tv("T")) {
-      case (r, interval) =>
-        val tinterval = TInterval(tv("T").t).physicalType
+      case (r, (intervalT: PInterval, interval)) =>
         val region = r.region
         val iv = r.mb.newLocal[Long]
         EmitTriplet(
-          Code(interval.setup, iv.storeAny(defaultValue(tinterval))),
-          interval.m || !Code(iv := interval.value[Long], tinterval.startDefined(region, iv)),
-          region.loadIRIntermediate(tv("T").t)(tinterval.startOffset(iv))
+          Code(interval.setup, iv.storeAny(defaultValue(intervalT))),
+          interval.m || !Code(iv := interval.value[Long], intervalT.startDefined(region, iv)),
+          region.loadIRIntermediate(tv("T").t)(intervalT.startOffset(iv))
         )
     }
 
     registerCodeWithMissingness("end", TInterval(tv("T")), tv("T")) {
-      case (r, interval) =>
-        val pinteval = TInterval(tv("T").t).physicalType
+      case (r, (intervalT: PInterval, interval)) =>
         val region = r.region
         val iv = r.mb.newLocal[Long]
         EmitTriplet(
-          Code(interval.setup, iv.storeAny(defaultValue(pinteval))),
-          interval.m || !Code(iv := interval.value[Long], pinteval.endDefined(region, iv)),
-          region.loadIRIntermediate(tv("T").t)(pinteval.endOffset(iv))
+          Code(interval.setup, iv.storeAny(defaultValue(intervalT))),
+          interval.m || !Code(iv := interval.value[Long], intervalT.endDefined(region, iv)),
+          region.loadIRIntermediate(tv("T").t)(intervalT.endOffset(iv))
         )
     }
 
     registerCode("includesStart", TInterval(tv("T")), TBooleanOptional) {
-      case (r, interval: Code[Long]) =>
-        PInterval(tv("T").t.physicalType).includeStart(r.region, interval)
+      case (r, (intervalT: PInterval, interval: Code[Long])) =>
+        intervalT.includeStart(r.region, interval)
     }
 
     registerCode("includesEnd", TInterval(tv("T")), TBooleanOptional) {
-      case (r, interval: Code[Long]) =>
-        PInterval(tv("T").t.physicalType).includeEnd(r.region, interval)
+      case (r, (intervalT: PInterval, interval: Code[Long])) =>
+        intervalT.includeEnd(r.region, interval)
     }
 
     registerCodeWithMissingness("contains", TInterval(tv("T")), tv("T"), TBoolean()) {
-      case (r, intTriplet, pointTriplet) =>
+      case (r, (intervalT: PInterval, intTriplet), (pointT, pointTriplet)) =>
         val pointType = tv("T").t.physicalType
```

Comment:
Is this different from pointT?

Reply:
fixed. The interval stuff is pretty broken regardless and will need more fixes before it's ready.

In particular the IRInterval class needs more than one ptype in its state, because the interval and point can be different.